### PR TITLE
MGMT-11536: refresh cluster after host deletion

### PR DESCRIPTION
**Pull request by slaviered**
When a user deletes a host, the cluster validations do not run immediately. Therefore the UI presents an incorrect view of the validation status. For example, master count validation shows erroneous status.

That may lead to a user starting an installation when the cluster is actually not ready to be installed.

This PR adds a refresh cluster after host deletion so the validation will run before the result is returned to the UI.

## List all the issues related to this PR

- [ ] New Feature 
- [x] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [x] Cloud
- [ ] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: subsystem/cluster_test.go [hunk 1]
```diff
@@ -218,6 +219,22 @@ var _ = Describe("Cluster", func() {
 		Expect(c.Hosts[0].ID.String()).Should(Equal(h.ID.String()))
 	})
 
+	It("Deregister host triggers cluster validations", func() {
```
Comment (nmagnezi) [inline subsystem/cluster_test.go:222]: Please also add a test for a negative case, where `refreshClusterStatus` returns an error.

File: internal/bminventory/inventory.go [hunk 2]
```diff
@@ -2872,6 +2872,9 @@ func (b *bareMetalInventory) V2DeregisterHostInternal(ctx context.Context, param
 		hostutil.GetHostnameForMsg(&h.Host))
 
 	if h.ClusterID != nil {
+		if _, err = b.refreshClusterStatus(ctx, h.ClusterID, b.db); err != nil {
```
Comment (nmagnezi) [inline internal/bminventory/inventory.go:2875]: Won't that make the flow synchronous?
Did you check how long it takes for `refreshClusterStatus` to run?

**Review by nmagnezi — CHANGES_REQUESTED** [review]

File: internal/bminventory/inventory.go (hunk 2 above)
Comment (slaviered) [reply internal/bminventory/inventory.go:2875]: This was exactly the intention of this PR. And yes this solution was approved. The Ux experience in SaaS takes precedence here.

File: subsystem/cluster_test.go (hunk 1 above)
Comment (slaviered) [reply subsystem/cluster_test.go:222]: It is not an interesting case since the update process just fails. What important here was to verify that the validations are returned synchronously.

Comment (nmagnezi) [reply subsystem/cluster_test.go:222]: I disagree.
Unit tests are meant to assert possible outcomes.
We should assert that if cluster sync fails, we return an error here.

Comment (slaviered) [reply subsystem/cluster_test.go:222]: Good point on both accounts. Adding a unit test

File: internal/bminventory/inventory.go [hunk 3]
```diff
@@ -5239,7 +5242,7 @@ func (b *bareMetalInventory) BindHostInternal(ctx context.Context, params instal
 	return host, nil
 }
 
-func (b *bareMetalInventory) UnbindHostInternal(ctx context.Context, params installer.UnbindHostParams, reclaimHost bool) (*common.Host, error) {
+func (b *bareMetalInventory) UnbindHostInternal(ctx context.Context, params installer.UnbindHostParams, reclaimHost bool, interactivity Interactivity) (*common.Host, error) {
```
Comment (ori-amizur) [inline internal/bminventory/inventory.go:5245]: The interactivity is not used

File: internal/bminventory/inventory.go [hunk 4]
```diff
@@ -2978,6 +2978,11 @@ func (b *bareMetalInventory) V2DeregisterHostInternal(ctx context.Context, param
 		hostutil.GetHostnameForMsg(&h.Host))
 
 	if h.ClusterID != nil {
+		if interactivity == Interactive {
+			if _, err = b.refreshClusterStatus(ctx, h.ClusterID, b.db); err != nil {
+				log.WithError(err).Warnf("Failed to refresh cluster after de-registerating host <%s>", params.HostID)
```
Comment (ori-amizur) [inline internal/bminventory/inventory.go:2983]: Only log ? Why not return error ?

Comment (slaviered) [reply internal/bminventory/inventory.go:2983]: Per discussion with @avishayt we decided to fail gracefully in such case (which should be rare).
